File: src/glasswyrmd/property_store.cpp
#include "glasswyrmd/resource_table.hpp"

#include <algorithm>
#include <new>

namespace glasswyrm::server {

PropertyMutationStatus ResourceTable::change_property(
    const std::uint32_t window_id, const std::uint32_t property_atom,
    Property value, const PropertyMode mode) {
  auto* window = find_window(window_id);
  if (window == nullptr) {
    return PropertyMutationStatus::BadWindow;
  }
  auto current = window->properties.find(property_atom);
  const bool exists = current != window->properties.end();
  if (exists && mode != PropertyMode::Replace &&
      (current->second.type != value.type ||
       current->second.format() != value.format())) {
    return PropertyMutationStatus::BadMatch;
  }

  try {
    Property replacement = std::move(value);
    if (exists && mode == PropertyMode::Append) {
      replacement.data = concatenate_property_data(current->second.data,
                                                   replacement.data);
    } else if (exists && mode == PropertyMode::Prepend) {
      replacement.data = concatenate_property_data(replacement.data,
                                                   current->second.data);
    }
    const std::size_t old_size = exists ? current->second.byte_size() : 0;
    const std::size_t new_size = replacement.byte_size();
    if (new_size > limits_.maximum_bytes_per_property ||
        (!exists && window->properties.size() >=
                        limits_.maximum_properties_per_window) ||
        new_size > limits_.maximum_total_property_bytes -
                       (total_property_bytes_ - old_size)) {
      return PropertyMutationStatus::BadAlloc;
    }

    if (exists) {
      current->second = std::move(replacement);
    } else {
      window->properties.emplace(property_atom, std::move(replacement));
    }
    total_property_bytes_ = total_property_bytes_ - old_size + new_size;
  } catch (const std::bad_alloc&) {
    return PropertyMutationStatus::BadAlloc;
  }
  return PropertyMutationStatus::Success;
}

bool ResourceTable::delete_property(const std::uint32_t window_id,
                                    const std::uint32_t property_atom) {
  auto* window = find_window(window_id);
  if (window == nullptr) {
    return false;
  }
  const auto iterator = window->properties.find(property_atom);
  if (iterator == window->properties.end()) {
    return true;
  }
  total_property_bytes_ -= iterator->second.byte_size();
  window->properties.erase(iterator);
  return true;
}

PropertyReadResult ResourceTable::get_property(
    const std::uint32_t window_id, const std::uint32_t property_atom,
    const std::uint32_t requested_type, const bool delete_after_read,
    const std::uint32_t long_offset, const std::uint32_t long_length) {
  auto* window = find_window(window_id);
  if (window == nullptr) {
    PropertyReadResult result;
    result.status = PropertyReadStatus::BadWindow;
    return result;
  }
  const auto iterator = window->properties.find(property_atom);
  if (iterator == window->properties.end()) {
    return {};
  }
  const auto& property = iterator->second;
  const std::size_t byte_size = property.byte_size();
  const std::uint64_t offset64 = static_cast<std::uint64_t>(long_offset) * 4U;
  if (offset64 > byte_size) {
    PropertyReadResult result;
    result.status = PropertyReadStatus::BadValue;
    return result;
  }

  PropertyReadResult result;
  result.present = true;
  result.value.type = property.type;
  result.value.format = property.format();
  result.type_matched = requested_type == 0 || requested_type == property.type;
  if (!result.type_matched) {
    result.value.bytes_after = static_cast<std::uint32_t>(byte_size);
    return result;
  }

  const std::size_t offset = static_cast<std::size_t>(offset64);
  const std::uint64_t requested64 = static_cast<std::uint64_t>(long_length) * 4U;
  const std::size_t available = byte_size - offset;
  const std::size_t returned = static_cast<std::size_t>(
      std::min<std::uint64_t>(available, requested64));
  result.value.bytes_after = static_cast<std::uint32_t>(available - returned);
  result.value.data = slice_property_data(property.data, offset, returned);
  if (delete_after_read && result.value.bytes_after == 0) {
    total_property_bytes_ -= byte_size;
    window->properties.erase(iterator);
    result.deleted = true;
  }
  return result;
}

std::vector<std::uint32_t> ResourceTable::list_properties(
    const std::uint32_t window_id) const {
  std::vector<std::uint32_t> result;
  const auto* window = find_window(window_id);
  if (window == nullptr) {
    return result;
  }
  result.reserve(window->properties.size());
  for (const auto& [atom, property] : window->properties) {
    static_cast<void>(property);
    result.push_back(atom);
  }
  std::sort(result.begin(), result.end());
  return result;
}

}  // namespace glasswyrm::server

File: crates/glasswyrm-core/tests/property_store.rs
use glasswyrm_core::property::{
    AtomId, Property, PropertyData, PropertyLimits, PropertyMode, PropertyMutationStatus,
    PropertyReadStatus,
};
use glasswyrm_core::resource_id::{ClientResourceRange, ResourceBase, ResourceMask};
use glasswyrm_core::window::{
    ClientId, CreateWindowStatus, WindowAttributes, WindowClass, WindowCreateSpec, WindowGeometry,
    WindowId, WindowStore,
};

const BASE: u32 = 0x0040_0000;
const WINDOW: WindowId = WindowId::new(BASE + 1);
const PROPERTY: AtomId = AtomId::new(39);
const STRING: AtomId = AtomId::new(31);

fn property(property_type: AtomId, data: PropertyData) -> Property {
    Property {
        property_type,
        data,
    }
}

fn store(limits: PropertyLimits) -> WindowStore {
    let mut store = WindowStore::new(Default::default(), limits);
    let status = store.create_window(
        ClientId::new(1),
        ClientResourceRange::new(ResourceBase::new(BASE), ResourceMask::new(0x001f_ffff)),
        WindowCreateSpec {
            xid: WINDOW,
            parent: WindowId::new(1),
            geometry: WindowGeometry {
                x: 0,
                y: 0,
                width: 10,
                height: 10,
                border_width: 0,
            },
            depth: 0,
            window_class: WindowClass::CopyFromParent,
            visual: 0,
            attribute_mask: 0,
            attributes: WindowAttributes::default(),
        },
    );
    assert_eq!(status, CreateWindowStatus::Success);
    store
}

#[test]
fn replace_append_and_prepend_preserve_type_format_and_order() {
    let mut store = store(PropertyLimits::default());
    let mutations = [
        (
            PropertyMode::Replace,
            PropertyData::U8(b"abc".to_vec()),
            b"abc".as_slice(),
        ),
        (
            PropertyMode::Append,
            PropertyData::U8(b"d".to_vec()),
            b"abcd".as_slice(),
        ),
        (
            PropertyMode::Prepend,
            PropertyData::U8(b"0".to_vec()),
            b"0abcd".as_slice(),
        ),
    ];
    for (mode, data, expected) in mutations {
        assert_eq!(
            store.change_property(WINDOW, PROPERTY, property(STRING, data), mode),
            PropertyMutationStatus::Success
        );
        assert_eq!(
            store
                .window(WINDOW)
                .unwrap()
                .property(PROPERTY)
                .unwrap()
                .data,
            PropertyData::U8(expected.to_vec())
        );
    }
    assert_eq!(store.total_property_bytes(), 5);

    let mismatches = [
        property(AtomId::new(19), PropertyData::U8(vec![1])),
        property(STRING, PropertyData::U16(vec![1])),
    ];
    for value in mismatches {
        assert_eq!(
            store.change_property(WINDOW, PROPERTY, value, PropertyMode::Append),
            PropertyMutationStatus::BadMatch
        );
    }
    assert_eq!(store.total_property_bytes(), 5);
    assert!(store.invariants_hold());
}

#[test]
fn reads_use_four_byte_units_and_delete_only_the_final_matching_slice() {
    let mut store = store(PropertyLimits::default());
    assert_eq!(
        store.change_property(
            WINDOW,
            PROPERTY,
            property(STRING, PropertyData::U8(b"0abcd".to_vec())),
            PropertyMode::Replace,
        ),
        PropertyMutationStatus::Success
    );

    let mismatch = store.get_property(WINDOW, PROPERTY, Some(AtomId::new(6)), true, 0, 10);
    assert_eq!(mismatch.status, PropertyReadStatus::Success);
    assert!(mismatch.present);
    assert!(!mismatch.type_matched);
    assert!(!mismatch.deleted);
    assert_eq!(mismatch.value.unwrap().bytes_after, 5);
    assert_eq!(store.total_property_bytes(), 5);

    let prefix = store.get_property(WINDOW, PROPERTY, Some(STRING), true, 0, 1);
    assert_eq!(
        prefix.value.unwrap().data,
        PropertyData::U8(b"0abc".to_vec())
    );
    assert!(!prefix.deleted);
    assert_eq!(store.total_property_bytes(), 5);

    let tail = store.get_property(WINDOW, PROPERTY, Some(STRING), true, 1, 10);
    let tail_value = tail.value.unwrap();
    assert_eq!(tail_value.property_type, STRING);
    assert_eq!(tail_value.format, 8);
    assert_eq!(tail_value.bytes_after, 0);
    assert_eq!(tail_value.data, PropertyData::U8(b"d".to_vec()));
    assert!(tail.deleted);
    assert_eq!(store.total_property_bytes(), 0);
    assert!(store.window(WINDOW).unwrap().property(PROPERTY).is_none());
}

#[test]
fn property_formats_slice_at_element_boundaries() {
    let cases = [
        (
            AtomId::new(40),
            property(AtomId::new(19), PropertyData::U16(vec![1, 2, 3])),
            16,
            PropertyData::U16(vec![3]),
        ),
        (
            AtomId::new(41),
            property(AtomId::new(6), PropertyData::U32(vec![1, 2, 3])),
            32,
            PropertyData::U32(vec![2]),
        ),
    ];
    for (atom, value, format, expected) in cases {
        let mut store = store(PropertyLimits::default());
        assert_eq!(
            store.change_property(WINDOW, atom, value, PropertyMode::Replace),
            PropertyMutationStatus::Success
        );
        let read = store.get_property(WINDOW, atom, None, false, 1, 1);
        let slice = read.value.unwrap();
        assert_eq!(slice.format, format);
        assert_eq!(slice.data, expected);
    }
}

#[test]
fn absent_invalid_offset_delete_and_sorted_listing_match_legacy_behavior() {
    let mut store = store(PropertyLimits::default());
    let absent = store.get_property(WINDOW, PROPERTY, None, false, 0, 1);
    assert_eq!(absent.status, PropertyReadStatus::Success);
    assert!(!absent.present);
    assert_eq!(
        store
            .get_property(WindowId::new(99), PROPERTY, None, false, 0, 1)
            .status,
        PropertyReadStatus::BadWindow
    );

    for atom in [AtomId::new(41), AtomId::new(39), AtomId::new(40)] {
        assert_eq!(
            store.change_property(
                WINDOW,
                atom,
                property(STRING, PropertyData::U8(vec![1, 2, 3])),
                PropertyMode::Replace,
            ),
            PropertyMutationStatus::Success
        );
    }
    assert_eq!(
        store.list_properties(WINDOW),
        [AtomId::new(39), AtomId::new(40), AtomId::new(41)]
    );
    assert!(store.list_properties(WindowId::new(99)).is_empty());
    let invalid = store.get_property(WINDOW, PROPERTY, None, true, 1, 1);
    assert_eq!(invalid.status, PropertyReadStatus::BadValue);
    assert!(store.window(WINDOW).unwrap().property(PROPERTY).is_some());
    assert!(store.delete_property(WINDOW, AtomId::new(40)));
    assert!(store.delete_property(WINDOW, AtomId::new(40)));
    assert!(!store.delete_property(WindowId::new(99), AtomId::new(40)));
    assert_eq!(store.total_property_bytes(), 6);
    assert!(store.invariants_hold());
}

#[test]
fn property_limits_reject_without_partial_mutation() {
    let mut store = store(PropertyLimits {
        maximum_bytes_per_property: 4,
        maximum_total_property_bytes: 6,
        maximum_properties_per_window: 2,
    });
    let cases = [
        (
            PROPERTY,
            PropertyData::U8(vec![1, 2, 3]),
            PropertyMode::Replace,
            PropertyMutationStatus::Success,
        ),
        (
            PROPERTY,
            PropertyData::U8(vec![4, 5]),
            PropertyMode::Append,
            PropertyMutationStatus::BadAlloc,
        ),
        (
            AtomId::new(40),
            PropertyData::U8(vec![4, 5, 6]),
            PropertyMode::Replace,
            PropertyMutationStatus::Success,
        ),
        (
            AtomId::new(41),
            PropertyData::U8(Vec::new()),
            PropertyMode::Replace,
            PropertyMutationStatus::BadAlloc,
        ),
        (
            AtomId::new(40),
            PropertyData::U8(vec![4, 5, 6, 7]),
            PropertyMode::Replace,
            PropertyMutationStatus::BadAlloc,
        ),
    ];
    for (atom, data, mode, expected) in cases {
        assert_eq!(
            store.change_property(WINDOW, atom, property(STRING, data), mode),
            expected
        );
    }
    assert_eq!(store.total_property_bytes(), 6);
    assert_eq!(
        store
            .window(WINDOW)
            .unwrap()
            .property(PROPERTY)
            .unwrap()
            .data,
        PropertyData::U8(vec![1, 2, 3])
    );
    assert_eq!(
        store
            .window(WINDOW)
            .unwrap()
            .property(AtomId::new(40))
            .unwrap()
            .data,
        PropertyData::U8(vec![4, 5, 6])
    );
    assert!(store.invariants_hold());
}
